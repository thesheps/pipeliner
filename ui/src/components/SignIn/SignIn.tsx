import React, { useState } from "react";
import Button from "@material-ui/core/Button";
import Dialog from "@material-ui/core/Dialog";
import DialogActions from "@material-ui/core/DialogActions";
import DialogContent from "@material-ui/core/DialogContent";
import DialogContentText from "@material-ui/core/DialogContentText";
import DialogTitle from "@material-ui/core/DialogTitle";
import TextField from "@material-ui/core/TextField";
import { connect } from "react-redux";

import { PipelinerState } from "../../store/types";
import { uiSelector } from "../../store/ui/selectors";
import { signInUserThunk } from "../../store/user/thunks";
import { setShowSignInModal } from "../../store/ui/actions";

export interface SignInProps {
  showSignInModal: boolean;
  setShowSignInModal: (show: boolean) => void;
  signInUser: (emailAddress: string, password: string) => void;
}

interface StateProps {
  showSignInModal: boolean;
}

export const SignIn = ({
  signInUser,
  setShowSignInModal,
  showSignInModal,
}: SignInProps) => {
  const [emailAddress, setEmailAddress] = useState("");
  const [emailError, setEmailError] = useState(false);
  const [password, setPassword] = useState("");

  const validateEmailAddress = (email: string) => {
    const re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    const isValid = re.test(email.toLowerCase());

    setEmailAddress(email);

    if (!isValid) {
      setEmailError(true);
    } else {
      setEmailError(false);
    }
  };

  return (
    <>
      <Button
        data-testid="signIn-dialog-button"
        color="inherit"
        onClick={() => setShowSignInModal(true)}
      >
        Sign In
      </Button>

      <Dialog open={showSignInModal} onClose={() => setShowSignInModal(false)}>
        <DialogTitle id="form-dialog-title">Sign In</DialogTitle>

        <DialogContent>
          <DialogContentText>Sign In here to get going!</DialogContentText>

          <TextField
            autoFocus
            value={emailAddress}
            onChange={(e) => validateEmailAddress(e.target.value)}
            inputProps={{ "data-testid": "emailAddress-input" }}
            margin="dense"
            label="Email"
            type="email"
            error={!!emailError}
            helperText="Please enter a valid email address"
            fullWidth
          />

          <TextField
            value={password}
            onChange={(e) => setPassword(e.target.value)}
            inputProps={{ "data-testid": "password-input" }}
            margin="dense"
            label="Password"
            type="password"
            fullWidth
          />
        </DialogContent>

        <DialogActions>
          <Button
            data-testid="cancel-button"
            onClick={() => setShowSignInModal(false)}
            color="primary"
          >
            Cancel
          </Button>

          <Button
            data-testid="signIn-button"
            onClick={() => signInUser(emailAddress, password)}
            color="primary"
            disabled={!emailAddress || !password || !!emailError}
          >
            Sign In
          </Button>
        </DialogActions>
      </Dialog>
    </>
  );
};

const mapStateToProps = (state: PipelinerState): StateProps => {
  const uiState = uiSelector(state);

  return {
    showSignInModal: uiState.showSignInModal,
  };
};

export const SignInContainer = connect(mapStateToProps, {
  signInUser: signInUserThunk,
  setShowSignInModal,
})(SignIn);
