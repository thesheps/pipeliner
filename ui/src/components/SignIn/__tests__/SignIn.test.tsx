import React from "react";
import { fireEvent, render } from "@testing-library/react";
import {
  toBeDisabled,
  toBeVisible,
  toHaveStyle,
} from "@testing-library/jest-dom/matchers";

import { SignIn } from "../SignIn";

expect.extend({ toBeDisabled, toBeVisible, toHaveStyle });

describe("SignIn", () => {
  const emailAddress = "joe.bloggs@invalid.com";
  const password = "my super awesome password";

  it("renders correctly", () => {
    const { container } = render(
      <SignIn
        showSignInModal={true}
        setShowSignInModal={jest.fn()}
        signInUser={jest.fn()}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it("calls setShowSignInModal on signIn click", () => {
    const setShowSignInModal = jest.fn();
    const { getByText } = render(
      <SignIn
        showSignInModal={false}
        setShowSignInModal={setShowSignInModal}
        signInUser={jest.fn()}
      />
    );

    fireEvent.click(getByText("Sign In"));

    expect(setShowSignInModal).toHaveBeenCalled();
  });

  it("validates emailAddress has been entered", () => {
    const { getByTestId } = render(
      <SignIn
        showSignInModal={true}
        setShowSignInModal={jest.fn()}
        signInUser={jest.fn()}
      />
    );

    fireEvent.change(getByTestId("password-input"), {
      target: { value: password },
    });

    expect(getByTestId("signIn-button")).toBeDisabled();
  });

  it("validates emailAddress is in the correct format", () => {
    const { getByTestId } = render(
      <SignIn
        showSignInModal={true}
        setShowSignInModal={jest.fn()}
        signInUser={jest.fn()}
      />
    );

    fireEvent.change(getByTestId("password-input"), {
      target: { value: password },
    });

    fireEvent.change(getByTestId("emailAddress-input"), {
      target: { value: "emailAddress" },
    });

    expect(getByTestId("signIn-button")).toBeDisabled();
  });

  it("validates password has been entered", () => {
    const { getByTestId } = render(
      <SignIn
        showSignInModal={true}
        setShowSignInModal={jest.fn()}
        signInUser={jest.fn()}
      />
    );

    fireEvent.change(getByTestId("emailAddress-input"), {
      target: { value: emailAddress },
    });

    expect(getByTestId("signIn-button")).toBeDisabled();
  });

  it("calls signInUser function upon clicking signIn", () => {
    const signInUser = jest.fn();
    const { getByTestId } = render(
      <SignIn
        showSignInModal={true}
        setShowSignInModal={jest.fn()}
        signInUser={signInUser}
      />
    );

    fireEvent.change(getByTestId("emailAddress-input"), {
      target: { value: emailAddress },
    });

    fireEvent.change(getByTestId("password-input"), {
      target: { value: password },
    });

    fireEvent.click(getByTestId("signIn-button"));
    expect(signInUser).toHaveBeenCalledWith(emailAddress, password);
  });

  it("calls hide modal function upon clicking cancel", () => {
    const setShowSignInModal = jest.fn();
    const { getByTestId } = render(
      <SignIn
        showSignInModal={true}
        setShowSignInModal={setShowSignInModal}
        signInUser={jest.fn()}
      />
    );

    fireEvent.click(getByTestId("cancel-button"));

    expect(setShowSignInModal).toHaveBeenCalled();
  });
});
