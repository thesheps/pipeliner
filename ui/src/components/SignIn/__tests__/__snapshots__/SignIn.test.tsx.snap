// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn renders correctly 1`] = `
<div
  aria-hidden="true"
>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
    data-testid="signIn-dialog-button"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      SignIn
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</div>
`;
